RevFileTree: Blame instead of View

gerhardol opened the pull request:
Part of #7598
Closes #9435
Closes #9437 

## Proposed changes

- RevFileTree: Blame instead of View
Show blame for object type blobs (files), as is available in FileHistory.
Note that object type commit (submodules) are shown as text, as before.
It is possible to toggle the view to show file contents, as before.

In FileHistory the commit info is available in a header. This is excluded in RevFileTree. The info for the selected commit can be seen in Commit, Diff tabs, overview is available in tooltip for otyher commits.
This could be configurable.

- RevDiff: HotKey to view FileTree
A menu command existed to view the file, this as a hotkey.

Note that `T` was selected as key, `F` is to be used to add the filepath to the Advanced filter in the grid.

- RevDiff: `B` to view Blame instead of Diff
Note that this is just for the current file, selecting another (or the same) file will show Diff by default.
For artificial, HEAD is blamed, worktree/index changes cannot be shown


- When blaming in RevDiff or FileTree, select the current line in fileviewer in Blame (especially useful in RevDiff)

- Cancellation support in LoadBlame() (in between operations, not in methods)

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/6248932/127392826-0f0da687-7ed4-46e6-b413-3ae19651da12.png)

For reference - FileHistory Blame tab, unchanged by this PR
![image](https://user-images.githubusercontent.com/6248932/127396308-6b386241-2228-4b1c-90e3-dbc4ede5dec9.png)

(RevDiff `B` opened FileHistory in a new instance, no screenshot.)

### After

![image](https://user-images.githubusercontent.com/6248932/127396404-ca3491e2-9329-45b6-acef-870b4e0591bb.png)

Only the hotkey added
![image](https://user-images.githubusercontent.com/6248932/127394925-5e676a30-3573-4627-a832-dbffb12d5e2f.png)

Context menu and hotkey `B` toggle Diff/Blame.

![image](https://user-images.githubusercontent.com/6248932/128437735-2b819e45-e440-4acb-8f86-ac2fec417ff2.png)

## Test methodology <!-- How did you ensure quality? -->

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -67,11 +69,14 @@ public RevisionFileTreeControl()
+        public void Bind(RevisionGridControl? revisionGrid, RevisionFileTreeControl revisionFileTree, Action refreshGitStatus)
         {
+            _revisionGrid = revisionGrid;
             _refreshGitStatus = refreshGitStatus;
+            _revisionFileTree = revisionFileTree;
```
Same sequence as arguments❔
```suggestion
            _revisionFileTree = revisionFileTree;
            _refreshGitStatus = refreshGitStatus;
```

gerhardol replied:
ok

mstv reviewed (commented):
👍 Feels quite natural.

Wishes for this PR:
- #9437 Update / paint FileTree independent from BlameControl in order to reduce flicker
- #9437 Make switching commits, e.g. `Blame previous revision`, not loose the file selection
![image](https://user-images.githubusercontent.com/36601201/127786020-1c9feb26-481c-4f0b-a0ca-86f120c1f149.png)
![image](https://user-images.githubusercontent.com/36601201/127786037-101e08b3-17a8-4860-9924-44346f186f46.png)

Wishes for later:
- #9437 Scroll FileTree selection to be visible when switching commits, e.g. on `Blame this / previous`
- #9436 Make Commit Info gutter of BlameControl collapsible (simulating "View"; improves scrolling UX)
- Store / restore splitter position of BlameControl (there is a pending issue for `master` anyway: 13caf92691cbefdf05cebb31454d7a8c0c7c3264)
  should be done in #9106
- #9435 Hide unrelated context menu item
![image](https://user-images.githubusercontent.com/36601201/127785914-9a3033a1-be2e-467f-94e9-0739e26c3d9c.png)
- #9434 BlameControl: Correctly align line numbers with text
![image](https://user-images.githubusercontent.com/36601201/127786108-1de106bc-f1cb-426c-aad8-8b30ca544446.png)

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -206,7 +206,7 @@ public FormBrowse(GitUICommands commands, string filter, ObjectId? selectedId =
             toolStripBranchFilterComboBox.DropDown += toolStripBranches_DropDown_ResizeDropDownWidth;
             revisionDiff.Bind(RevisionGrid, fileTree, () => RequestRefresh());
-            fileTree.Bind(() => RequestRefresh());
+            fileTree.Bind(RevisionGrid, fileTree, () => RequestRefresh());
```
ok

gerhardol commented on GitUI/CommandsDialogs/RevisionFileTreeControl.Designer.cs:
```diff
@@ -360,6 +371,7 @@ private void InitializeComponent()
         private System.Windows.Forms.SplitContainer FileTreeSplitContainer;
         private UserControls.NativeTreeView tvGitTree;
         private Editor.FileViewer FileText;
+        private Blame.BlameControl blameControl;
```
will change

gerhardol commented:
> Wishes for this PR:

Most of these are for BlameControl should be a separate issue (#7598 is quite confusing as is...)
Improvements are necessary.

Also if Blame is improved, there may be need for a configurable way to  show Blame/View (or even Blame/Diff).
Some discussions in #7598, this is a first implementation to fond what is really needed.

gerhardol commented:
Converted this to a draft and based it on #9425 (approved, to be merged soon) and #9426 as there are some functionality required.
The commits from 7037db268d7d61aa3fdffe75f4469cc60321eaf0 *RevFileTree: Blame instead of View* to 1c921d7f27c8e20b7835dfb3ad3450ce5c26b825 *fixup! RevFileTree: Blame instead of View* are unchanged.

Some new additions:
* flicker selecting is removed in c7f0165d0ed6b51077d552cd0ff3c3dd476cdbf0 *fixup! avoid flicker switching files*
* The existing issue is removed 58e63c02e5ac9697c766cb292d5177bd870b975b *Blame; Hide EnableAutomaticContinuousScroll*
* To avoid that slow blame is a blocker, it is possible to View files as before.
The shortcut key `B` is used to toggle View/Blame instead of opening a new instance. (the value is stored in settings.
Note: a similar change should be added to RevDiff, `B` to Blame files.

gerhardol commented:
Cleaned up commits a little, adding RevDiff Blame
Ready for review again
Further changes are presented in #9445

GintasS commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -51,26 +51,29 @@ public partial class RevisionFileTreeControl : GitModuleControl
+        public void Bind(RevisionGridControl? revisionGrid, Action refreshGitStatus)
         {
+            _revisionGrid = revisionGrid;
             _refreshGitStatus = refreshGitStatus;
```
Since you're using nullable reference type for `RevisionGridControl`, I'm wondering if the `Action refreshGitStatus `parameter should be too `Action? refreshGitStatus` here?

gerhardol replied:
That depends on what the caller provides, but in this case it should be changed
(comments on existing code is welcome, bigger changes should be separated commits or even PRs though)

gerhardol replied:
`RevisionGridControl` should not be NRT, caller is always non null
In RevDiff `refreshGitStatus` should be NRT too.

GintasS commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -351,35 +356,59 @@ private void tvGitTree_AfterSelect(object sender, TreeViewEventArgs e)
+                case GitObjectType.Blob:
                     {
-                        GitItemStatus file = new(name: gitItem.FileName)
+                        if (!blameToolStripMenuItem1.Checked)
```
What do you think about not using this logical negation and instead writing `blameToolStripMenuItem1.Checked == false`? Would be more readable, you can miss that `!`

GintasS replied:
Also, in the method `ExecuteCommand` in `RevisionDiffControl,cs`, you use (reuse) one style of switch statements (making cases inline) and here you do in one style, FYI.

gerhardol replied:
> What do you think about not using this logical negation and instead writing blameToolStripMenuItem1.Checked == false? Would be more readable, you can miss that !

In this case, I prefer the compact style

> Also, in the method ExecuteCommand in RevisionDiffControl,cs, you use (reuse) one style of switch statements (making cases inline) and here you do in one style, FYI.

Is this about how Blame is activated?
In that case it is intentional, FileTree remembers Blame/View for those that prefer to only view (that is faster).
For RevDiff it is always a one time activation, next selection is always Diff (or New etc) again.

RussKie replied:
> What do you think about not using this logical negation and instead writing `blameToolStripMenuItem1.Checked == false`? Would be more readable, you can miss that `!`

This is C# intrinsic (and C-like languages in general). The only time one would write ` == false` when the left operand is a nullable.

GintasS commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -351,35 +356,59 @@ private void tvGitTree_AfterSelect(object sender, TreeViewEventArgs e)
+
+            Task ViewGitItemAsync(GitItem gitItem)
+            {
+                GitItemStatus file = new(name: gitItem.FileName)
```
Just out of curiosity, what is this syntax, this is C# 9.0 feature?

gerhardol replied:
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9#fit-and-finish-features
Compact code - clearer and more compact than `var`

GintasS commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -428,10 +457,15 @@ private void tvGitTree_KeyDown(object sender, KeyEventArgs e)
+                return;
             }
+
+            blameToolStripMenuItem1.Checked = !blameToolStripMenuItem1.Checked;
```
You can miss that `!` so easily here, but it is up to you.

gerhardol replied:
I guess pattern matching could be used, but I want to have `!` as our convention

GintasS commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -212,6 +215,8 @@ public void LoadRevision(GitRevision? revision)
                 {
+                    BlameControl.Visible = false;
+                    FileText.Visible = true;
                     FileText.Clear();
```
What do you think about having a separate method to control these `BlameControl `and `FileText `Visible states? I can see in a couple of instances you setting these two to false and true respectively.

gerhardol replied:
Similar in FileTree and RevDiff, just one of the controls are active at a time.
In this case there are only two controls and only true/false switch so I prefer to not hide this in a method.
For RevDiff, I separated `BlameSelectedFileDiffAsync()` to a separate method that was kept close to the existing `ShowSelectedFileDiffAsync()` to show the similarities.
(Good suggestion still...)

GintasS reviewed (commented):
Wrote a few comments regarding style basically.

gerhardol commented:
Rebased on master to resolve MC (and master eeb8f67ddbf87ccc88741fe31c31ac444948a266 is unusable, fixed in adbb410d5b65b8d23d8bb71541179cbe573478e8).
Also squashed the NRT in https://github.com/gitextensions/gitextensions/pull/9424#discussion_r685302418

mstv commented on GitUI/CommandsDialogs/RevisionFileTreeControl.Designer.cs:
```diff
@@ -339,6 +341,15 @@ private void InitializeComponent()
+            this.BlameControl.Margin = new System.Windows.Forms.Padding(0);
+            this.BlameControl.Name = "blameControl";
+            this.BlameControl.Size = new System.Drawing.Size(487, 303);
+            this.BlameControl.TabIndex = 0;
```
Conflict with `FileText.TabIndex` (although they are not displayed at the same time).

gerhardol replied:
That was intentional, really a need to separate them?

RussKie replied:
As a general rule: no item should have identical indices. Though this is the keyboard tab order, i.e. used when tab between controls. Accessibility tools don't use it (AFAIK), and use the order in which controls are added into respective parents instead.

gerhardol replied:
Do you request to change this?

mstv reviewed (commented):
LGTM

gerhardol commented:
Rebased without changes, to simplify reviews and get the build successful

RussKie commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -51,26 +51,29 @@ public partial class RevisionFileTreeControl : GitModuleControl
         private readonly IRevisionFileTreeController _revisionFileTreeController;
         private readonly IFullPathResolver _fullPathResolver;
-        private readonly IFindFilePredicateProvider _findFilePredicateProvider;
+        private readonly IFindFilePredicateProvider _findFilePredicateProvider = new FindFilePredicateProvider();
```
Why is this necessary?

gerhardol replied:
It is not necessary for this PR, it was a change I had done (and believe should be changed) that was refactored to a separate commit

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -449,8 +453,21 @@ private void ResetSelectedItemsTo(bool actsAsChild)
             RequestRefresh();
         }
 
+        private Task BlameSelectedFileDiffAsync()
+        {
+            BlameControl.Visible = true;
+            DiffText.Visible = false;
+            GitRevision rev = DiffFiles.SelectedItem.SecondRevision.IsArtificial
+                ? _revisionGrid.GetActualRevision(_revisionGrid.CurrentCheckout)
+                : DiffFiles.SelectedItem.SecondRevision;
+            BlameControl.LoadBlame(rev, null, DiffFiles.SelectedItem.Item.Name, _revisionGrid, null, DiffText.Encoding);
+            return Task.CompletedTask;
+        }
```
This is really confusing - there is nothing async in this method, and everything must be done on the UI thread.
Please change it into a sync method, and update the the call site to invoke it async, if necessary.

gerhardol replied:
This was done in the same way as `ShowSelectedFileDiffAsync()`, the call in `blameToolStripMenuItem_Click()` need to async anyway (even if this is sync).
Cancelation token should be added to LoadBlame too, similar to how `ViewChangesAsync()` was changed.
(that will require Git commands to be cancellable first to be really useful)

So I would prefer to keep it as is.

gerhardol replied:
Changed this - there are operations blocking the UI thread, but not for long.

RussKie commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -351,35 +356,59 @@ private void tvGitTree_AfterSelect(object sender, TreeViewEventArgs e)
-                        return FileText.ViewGitItemAsync(file, gitItem.ObjectId);
+                        BlameControl.Visible = true;
+                        FileText.Visible = false;
+                        BlameControl.LoadBlame(_revision, null, gitItem.FileName, _revisionGrid, null, FileText.Encoding);
```
Please add argument names for `null`.

gerhardol commented:
Added two more features:
* When blaming in RevDiff or FileTree, select the current line in fileviewer in Blame (especially useful in RevDiff)
* Cancellation support in LoadBlame() (in between operations, not in methods)

RussKie commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -51,26 +51,29 @@ public partial class RevisionFileTreeControl : GitModuleControl
-        public void Bind(Action refreshGitStatus)
+        public void Bind(RevisionGridControl revisionGrid, Action? refreshGitStatus)
         {
+            _revisionGrid = revisionGrid;
```
I very much dislike this - this creates further coupling between various controls, and making it harder to reason about each individual control's functionality, and further complicating testing stories.

gerhardol replied:
This is required for BlameControl so the same as in FormFileHistory.
You can see this as a step to remove a lot of code in GE: The follow up #9445 obsoletes FormFileHistory, so this is a step to remove a complete form.

gerhardol commented:
Rebased and squashed without changes, prepared for rebase merge when approved.

gerhardol commented:
Rebased to resolve a merge conflict

RussKie commented:
I ran it locally briefly, seems to work. Need time to form a better opinion, but overall I think it is a good change.
Two things I've noticed:
1. "Blame" isn't sticky, i.e. when change file selection the view changes to the "diff". Feels weird.
2. Attempt to open history "feels" like it getting ignored. At least there is no immediate feedback, the new window doesn't open, or opens after a significant delay.

gerhardol commented:
>     1. "Blame" isn't sticky, i.e. when change file selection the view changes to the "diff". Feels weird.

Blame/View is sticky in FileTree tab.
Intentional in diff tab, I want to go to diff quickly in diff tab. This is partly as BlameControl is slower, annoying to show first then to blame.
I have used this for a month or so now, this is something I appreciate so I prefer to keep this.

>     2. Attempt to open history "feels" like it getting ignored. At least there is no immediate feedback, the new window doesn't open, or opens after a significant delay.

Do you mean H? That is not changed by this PR.
Or open FileTab? Open FileTree with Blame requires a little longer time to open Blame to View, That is why Blame/View is sticky, stored in settings.
